### ISSUE TITLE
Adding MeshCNN to PyTorch Geometric

### ISSUE THREAD

**Issue by ranahanocka** [rusty1s/pytorch_geometric#283]
It would be so awesome if you could add MeshCNN to PyTorch Geometric. We have just released the [code](https://github.com/ranahanocka/MeshCNN) (PyTorch)

**Comment by rusty1s** [comment]
Thanks for letting me know. I see what I can do, although I would appreciate a pull request if you find the time.

**Comment by jlevy44** [comment]
I second this! Nice repo @ranahanocka !

If I had more time, I'd help out!

**Comment by omria503** [comment]
Hi,
I'm trying to implement MeshCNN with PyTorch Geometric. 
Since I'm new with PyTorch Geometric and the basic stuff of graphs I would appreciate your help and comments.
I would like to implement the first layer which is the mesh convolution:
According to the [MeshCNN paper](https://arxiv.org/pdf/1809.05910.pdf):
![image](https://user-images.githubusercontent.com/60508587/101246053-7d402d00-3719-11eb-829f-6df91d94ea48.png)

One important note is that each edge holds a feature vector, not each node, therefore we can think of it as the edges are the nodes in a dual graph:
![image](https://user-images.githubusercontent.com/60508587/101246143-1a02ca80-371a-11eb-8201-517249ed4df4.png)

I've planned to implement it as the following:
edge_index - a 2d Tensor with shape of 2xE, it holds all the edge connections to each other.
In the example above: [[0, 0, 0, 0, 0, 1, 2, 3, 4, ...], [0, 1, 2, 3, 4, 0, 0, 0, 0, ...]]
x is a tensor with all the features for all the connections.
In the message function I've created the actual features (e, |a-c|, a+c, |b-d|, b+d) and used nn.Linear layer instead of convolution layer. Next, an aggregation function applied - "add". Last, the update function will return the aggregation output.

Do you think this is the correct way (using linear layer + aggregation "add" instead of convolution)?

Thanks

**Comment by rusty1s** [comment]
Yes, `Linear` layer and "sum" aggregation is the way to go to implement MeshCNN IMO. However, as far as I can see, edge indices needs to be sorted to guarantee correct computation of `e_1`, `e_2`, ...

**Comment by Omri-L** [comment]
Hi,
I am planning to send a pull request in one of the next weeks for [MeshCNN ](https://arxiv.org/pdf/1809.05910.pdf)layers.
In order to make it easier for me, I will appreciate if you could advise about the following modules:

1. **MeshConv** - implements the mesh convolution layer according to the paper. It uses MessagePassing pytroch geometric module.
   Should be in "pytorch_geometric\torch_geometric\nn\conv\mesh_conv.py" (?)
2. **Mesh** class - a data structure which reads the data from an obj file and creates relevant data for internal use (see below modules). 
3. **MeshPrepapre** class - reads an obj file, prepares the data for Mesh class such as extracting mesh features and creates mesh augmentations.
4. **MeshUnion** class - a utility class which unifies mesh edges and features.
5. **MeshPool** - implements the mesh pooling layer according to the paper. It uses a utility class named MeshUnion and a the structure class named Mesh.
   Should be in "pytorch_geometric\torch_geometric\nn\pool\mesh_pool.py" (?)
6. **MeshUnpool** - implements the mesh unpooling layer according to the paper. It uses Mesh class.
   Should be in "pytorch_geometric\torch_geometric\nn\unpool\mesh_pool.py" (?)

Mesh augmentations, Mesh features and Mesh utils should be on different files but are not classes.

**My questions are:**
1. Where should I put each class or file?
2. For which class should I add a unit test?
3. Should I create a full example of use of loading data and building a small network which uses the above modules? (such as in MeshCNN [github](https://github.com/ranahanocka/MeshCNN)).


Thanks!!!

**Comment by rusty1s** [comment]
Thanks for your interest in contributing.

1. Do we really need an own `Mesh` class or is `Data` sufficient?
2. I guess `MeshCNNPrepare` should be implemented as a transform in `torch_geometric.transforms`.

It would be great to have a unit test for `MeshConv`, `MeshPrepare` and `MeshPool`, but those are not mandatory. We can create them afterwards.
A full example to show-case its usage would be awesome :)

**Comment by Omri-L** [comment]
1. Mesh class - I did not described the class functionality as it is. It is not just a data structure but it also have some functionality (such as adding or removing edges, vertexes, exporting data to file and more). In addition, it holds more fields (but I saw that Data can be extended).
2. I will look into it.

Thanks

**Comment by sharmishtaa** [comment]
I support this idea! I would love to see MeshCNN here. Are there any updates on when we might be merging?

**Comment by gattia** [comment]
@Omri-L

I notice you have a fork w/ some progress on mesh cnn. How close is that to functioning?

**Comment by daniel-unyi-42** [comment]
@Omri-L @rusty1s any updates on this? I think MeshCNN would be an important addition to PyG and offer my help to finish things up.

**Comment by tpprytula** [comment]
I support the idea!! It is been a couple of times now that I wanted to use MeshCNN, but I always ended up in resorting to some less fitting GNN architectures just because they're available in PyG.

**Comment by michaelfortunato** [comment]
Hi everyone. I am so excited to try to work on this. I believe MeshCNN is a very important paper and would fit quite nicely with PyTorch Geometric. Here, I give a summary of my proposal to you all to integrate MeshCNN into PyTorch Geometric, but the full plan and a comprehensive explanation of MeshCNN is here [MeshCNN-PyG.pdf](https://github.com/user-attachments/files/19872710/MeshCNN-PyG.pdf)

>[!NOTE]
> I am a student in Chicago and have the time necessary to dedicate myself to this effort for the next few months.  If you have thoughts, questions, suggestions and contributions, please feel more than free to reach out to me directly on Pytorch Geometric's Slack Channel. I would love to learn from each other, answer questions, and/or collaborate together. My display name  `@Michael Fortunato` and my profile URL is <https://torchgeometricco.slack.com/team/U08AMAGE318>. I am always online and more than happy to hop on a video or voice call to discuss MeshCNN.

 
# Background 

>[!IMPORTANT]
> Please read the full plan: [MeshCNN-PyG.pdf](https://github.com/user-attachments/files/19872710/MeshCNN-PyG.pdf). This comment is only a synopsis for convenience.

The MeshCNN neural network can best be viewed as a function on triangular meshes 
$m = (V, F) \in \mathbb{R}^{|V| \times 3} \times [0, ..., |V| - 1\]^{3 \times |F|}$, 
that given a set of categories $C$, returns a $C$-dimensional vector, whose $i$ th component denotes the probability of the input mesh $m$ belonging to category $c_i \in C$.   MeshCNN has 5 layers, which are functions that operate on the outputs of the prior layer.

- Layer 1: _The feature extraction layer_ ("Feat").This layer is a function that takes as input the triangular mesh $m$ and returns a $|E| \times 5$ edge feature matrix along with the edge adjacency of the input mesh $m$. Please see the full PDF for further explanation. 
- Layer 2: _The convolutional layer_("Conv"). This layer takes as input the set of edge features computed by the prior layer $X^{(k)} \in \mathbb{R}^{|E| \times d(k)}$
  along with the original edge adjacency matrix of the input mesh $m$ and returns an updated edge feature $X^{(k+1)} \in \mathbb{R}^{|E| \times d(k+1)}$ to be fed into subsequent layers. Please see the original paper or the attached PDF for how this layer computes $X^{(k+1)}$. A good overview may be found [here](https://pytorch-geometric--10222.org.readthedocs.build/en/10222/generated/torch_geometric.nn.conv.MeshCNNConv.html)
- Layer 3: _The edge pooling layer_ ("Pool"): Pools edges that leave the mesh manifold, in the order of some priority. 
- Layer 4: _The unpooling layer_ ("UnPool"): Restores the original topology of the mesh, but masks the edges with the features computed in Layer 3 over the pooled edges.
- Layer 5: _The probability layer_ ("Prob"): Returns a $|C|$-dimensional vector, call it $a$, whose components sum to 1. $a_i$ denotes the probability that input mesh $m$ belongs to category $c_i \in C$.

# The Proposal To Add MeshCNN Into PyTorch Geometric.    



- [x] Part 1: Add MeshCNN's convolutional layer (Layer 2) to PyTorch Geometric
   - `torch_geometric/nn/conv/meshcnn_conv.py`: The convolutional layer
   - `test/nn/conv/test_meshcnn_conv.py`: Unit test class 
   - The PR: https://github.com/pyg-team/pytorch_geometric/pull/10222
- [ ] Part 2: Add MeshCNN's feature extraction (Layer 1) layer to PyTorch Geometric
   - `torch_geometric/nn/models/meshcnn.py` - The class stub holding the MeshCNN model. Layer 1 will be implemented here as a private method, as it is specific to MeshCNN. Any usage of this class before this issue is closed will return TODO. 
   - `torch_geometric/utils/mesh_geometry.py`: This class will compute the geometric features associated with each edge that are used by layer 1 of MeshCNN. It utility is general enough to warrant a separate module for it so that it may be used in other scenarios.   
   - The PR: TBD
- [ ] Part 3: Add MeshCNN's pooling and unpooling layer (possibly a not as a separate layer) to PyTorch Geometric. Finishing the integration. My plan for it is TBD until we get the first two pull requests in.
    - The PR: TBD




# Acknowledgements

- It's important to recognize that the original pull request offering to integrate MeshCNN into PyG, https://github.com/pyg-team/pytorch_geometric/pull/2293, is quite comprehensive and complete. The code we put here for MeshCNN will probably look very much like the original PR made https://github.com/pyg-team/pytorch_geometric/pull/2293. The only real difference is that we will split up the pull requests into a series of smaller ones, which will be heavily inspired by the ideas from https://github.com/pyg-team/pytorch_geometric/pull/2293.
- A link to the original MeshCNN paper: https://arxiv.org/abs/1809.05910


Thank you and please let me know what you think. Once again, I am always online on PyG's Slack channel, and encourage reviewers to reach out to me directly there over message, whatever is most convenient. 

Thanks so much,
Michael

Slack: `@Michael Fortunato` on [Slack](https://torchgeometricco.slack.com/team/U08AMAGE318)
Plan: [MeshCNN-PyG.pdf](https://github.com/user-attachments/files/19872710/MeshCNN-PyG.pdf)
PR 1: <https://github.com/pyg-team/pytorch_geometric/pull/10222>

**Comment by michaelfortunato** [comment]
I am close to getting Part 2 PR ready to post. I have two implementations of Part 2, one I wrote by hand, and the other using Trimesh. That made me think that ideally, it would be nice if I could identify the primitive operations needed to compute edge features, and then allow users to provide a "backend" implementing those necessary trait methods--essentially employ. the strategy pattern/dependency injection.  In a similar vein, I am starting to think about mesh pooling. This may be jumping ahead a bit, but there may be an opportunity to perform edge collapse on the GPU.  In general, it might be nice is MeshCNN's pooling layer required that a `Data` object implement a "Contractible" trait, something like 
```python
def can_contract_edge(self, edge_id):
   r"""
   returns true if contracting edge edge_id leaves self (the mesh) manifold
   """
   ...
def contract_edge(self, edge_id):
    r"""
    Contracts the edge regardless in self regardless if result is manifold. 
    Returns 'decimated_edge' i.e. any context necessary to restoring the original mesh
    """
def restore(self, decimated_edges):
  r""" restores the original topology of the mesh given the recovery information 'decimated edges'.  """
```

Letting a consumer of MeshCNN's pooling layer provide a "mesh backend" would be great for the model. Then, if say a user did not want a third party dependency it would default to our implementation of edge collapse on the CPU in Python that we would write. Otherwise, if they provide a backend, they could perform edge collapse on the GPU. Will be very curious to see if there is any difference in speed. 

Here is an implementation of edge collapse on a GPU. 
https://github.com/owensgroup/RXMesh/blob/main/apps/SurfaceTracking/collapser.cuh
The kernel is around 300 lines. I'll need to look at it to see if it in fact could be used here.

**Comment by michaelfortunato** [comment]
*Update (07/15/25)*: For anyone waiting for the full model. I apologize for the delay. I am quite busy this summer, and expect to get back to finishing up the integration in late September. Hopefully its clear that Part 2 is essentially ready for PR. I will keep referencing this issue in subsequent commits so those who are interested can track progress in this issue. 

Thanks,
Michael

--- 

Three updates. 
1. I have created a faster computation of the edge adjacency matrix than what was used in the original MeshCNN repo. I see significant speed up and GPU usage (I have access to a 3080 consumer GPU)  on my machine as compared to the original repo.
2. In terms of edge collapsing (PR 3), I started speaking with the creator of RXMesh he gave tips for using his collapser kernel, but also of course gave the general advice said that if we can find areas to remove any loops or branching, we might be able to get a lot of speed up. In fact, I found this to be true with the feature extraction (PR 2), so glad I did not try to write a kernel for that! 
3. **I've had PR 2 (the feature extraction layer) ready on my machine, so given PR 1 is now done,  I will put that up this hopefully by this weekend (I have some deadlines this week).** 

After that I will put the final part, PR 3 up for review. 
Assuming all goes well that will complete the integration!

---

For those interested, here is how we will compute the adjacency. I'm quite happy with its results. Note that both this implementation and MeshCNN's original implementation implicitly rely on the fact the mesh's winding is encoded in the data. I will elaborate more on what winding is in comments in the code and maybe add a function to check winding. For now, its enough to note that winding asserts the nodes in F of a M=(V, F) mesh are ordered geometrically (ccw or cw). See here https://trimesh.org/trimesh.base.html#trimesh.base.Trimesh.is_winding_consistent for an explanation now.



```python

def adj(face: torch.Tensor):
    edges = torch.stack(
        [
            face[[0, 1], :],  # v0 -> v1
            face[[1, 2], :],  # v1 -> v2
            face[[2, 0], :],  # v2 -> v0
        ],
        dim=2).reshape(2, -1)

    sorted_edges, _ = torch.sort(edges, dim=0)

    unique_edges, edge_ids, edge_counts = torch.unique(
        sorted_edges, dim=1, return_inverse=True, return_counts=True)

    face_edges = edge_ids.view(-1, 3)

    neighbor_pairs = torch.stack(
        [face_edges[:, [(i + 1) % 3, (i + 2) % 3]] for i in range(3)],
        dim=1).reshape(-1, 2)

    sorted_edge_ids, order = torch.sort(edge_ids)
    sorted_neighbors = neighbor_pairs[order]

    first_occurrence_mask = torch.cat([
        torch.tensor([True], device=edges.device), sorted_edge_ids[1:]
        != sorted_edge_ids[:-1]
    ])

    first_positions = first_occurrence_mask.nonzero(as_tuple=True)[0]

    adjacency = torch.full((unique_edges.size(1), 4), -1,
                           dtype=torch.long, device=edges.device)

    adjacency[:, :2] = sorted_neighbors[first_positions]

    interior_edge_indices = (edge_counts == 2).nonzero(
        as_tuple=True)[0]
    second_positions = first_positions[interior_edge_indices] + 1
    adjacency[interior_edge_indices,
              2:] = sorted_neighbors[second_positions]

    return adjacency
``` 

Compare this to  the original implementation. It uses for loops and is more readable, but is slower on my machine (both for my CPU or GPU):
https://github.com/ranahanocka/MeshCNN/blob/5bf0b899d48eb204b9b73bc1af381be20f4d7df1/models/layers/mesh_prepare.py#L116-L161
